// near-api-js/lib/rpc/types.gen.js
// This file is auto-generated by @hey-api/openapi-ts
export const Direction = { LEFT: 'Left', RIGHT: 'Right' };
/**
 * Different types of finality.
 */
export const Finality = {
    OPTIMISTIC: 'optimistic',
    NEAR_FINAL: 'near-final',
    FINAL: 'final',
};
export const GenesisConfigError = { NULL: null };
/**
 * GenesisConfigRequest
 */
export const GenesisConfigRequest = { NULL: null };
export const LogSummaryStyle = { PLAIN: 'plain', COLORED: 'colored' };
export const MethodResolveError = {
    METHOD_EMPTY_NAME: 'MethodEmptyName',
    METHOD_NOT_FOUND: 'MethodNotFound',
    METHOD_INVALID_SIGNATURE: 'MethodInvalidSignature',
};
/**
 * Configures whether the node checks the next or the next next epoch for network version compatibility.
 */
export const ProtocolVersionCheckConfig = { NEXT: 'Next', NEXT_NEXT: 'NextNext' };
/**
 * RpcClientConfigRequest
 */
export const RpcClientConfigRequest = { NULL: null };
/**
 * RpcHealthRequest
 */
export const RpcHealthRequest = { NULL: null };
export const RpcHealthResponse = { NULL: null };
/**
 * RpcNetworkInfoRequest
 */
export const RpcNetworkInfoRequest = { NULL: null };
/**
 * RpcStatusRequest
 */
export const RpcStatusRequest = { NULL: null };
/**
 * This enum represents if a storage_get call will be performed through flat storage or trie
 */
export const StorageGetMode = { FLAT_STORAGE: 'FlatStorage', TRIE: 'Trie' };
export const SyncCheckpoint = { GENESIS: 'genesis', EARLIEST_AVAILABLE: 'earliest_available' };
