// near-api-js/lib/accounts/nonce-manager.js
export class NonceManager {
    constructor() {
        this.nonces = {};
        this.locks = {};
    }
    async resolveNextNonce(pk, accountId, provider) {
        const key = pk.toString();
        // wait until unlocked
        const lock = this.locks[key];
        if (lock)
            await lock;
        if (typeof this.nonces[key] === 'bigint') {
            this.nonces[key] = this.nonces[key] + 1n;
            return this.nonces[key];
        }
        // biome-ignore lint/suspicious/noEmptyBlockStatements: unlock is immediately overwritten
        let unlock = () => { };
        const newLock = new Promise((resolve) => {
            unlock = resolve;
        });
        this.locks[key] = newLock;
        try {
            const accessKey = await provider.viewAccessKey({
                accountId,
                publicKey: pk,
                finalityQuery: { finality: 'optimistic' },
            });
            this.nonces[key] = accessKey.nonce + 1n;
            return this.nonces[key];
        }
        finally {
            unlock();
        }
    }
    async invalidate(pk) {
        const key = pk.toString();
        // make sure to wait until unlocked
        const lock = this.locks[key];
        if (lock)
            await lock;
        delete this.nonces[key];
    }
}
