// near-api-js/lib/providers/errors/contract_execution.js
import { HandlerError } from './handler.js';
export class ContractExecutionError extends HandlerError {
    constructor(message, blockHash, blockHeight) {
        super(message);
        this.blockHash = blockHash;
        this.blockHeight = blockHeight;
    }
}
export class WasmTrapError extends ContractExecutionError {
    constructor(reason, blockHash, blockHeight) {
        super(`WASM trap during contract execution at block height ${blockHeight}`, blockHash, blockHeight);
        this.reason = reason;
    }
}
export class ContractCodeDoesNotExistError extends ContractExecutionError {
    constructor(contractId, blockHash, blockHeight) {
        super(`Contract code for contract ID ${contractId} does not exist at block height ${blockHeight}`, blockHash, blockHeight);
        this.contractId = contractId;
    }
}
export class ContractMethodNotFoundError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Contract method is not found at block height ${blockHeight}`, blockHash, blockHeight);
    }
}
export class ContractMethodNameEmptyError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Contract method name is empty`, blockHash, blockHeight);
    }
}
export class ContractMethodInvalidSignatureError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Contract method invalid signature`, blockHash, blockHeight);
    }
}
export class PrepareWasmError extends ContractExecutionError {
    constructor(reason, blockHash, blockHeight) {
        super(`Failed to prepare WASM because of '${reason}'`, blockHash, blockHeight);
        this.reason = reason;
    }
}
export class HostBadUtf16Error extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Bad UTF-16', blockHash, blockHeight);
    }
}
export class HostBadUtf8Error extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Bad UTF-8', blockHash, blockHeight);
    }
}
export class HostGasExceededError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Gas exceeded', blockHash, blockHeight);
    }
}
export class HostGasLimitExceededError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Gas limit exceeded', blockHash, blockHeight);
    }
}
export class HostBalanceExceededError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Balance exceeded', blockHash, blockHeight);
    }
}
export class HostEmptyMethodNameError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Empty method name', blockHash, blockHeight);
    }
}
export class HostGuestPanicError extends ContractExecutionError {
    constructor(panicMessage, blockHash, blockHeight) {
        super(`Contract panicked: ${panicMessage}`, blockHash, blockHeight);
        this.panicMessage = panicMessage;
    }
}
export class HostIntegerOverflowError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Integer overflow', blockHash, blockHeight);
    }
}
export class HostInvalidPromiseIndexError extends ContractExecutionError {
    constructor(promiseIndex, blockHash, blockHeight) {
        super(`Invalid promise index ${promiseIndex}`, blockHash, blockHeight);
        this.promiseIndex = promiseIndex;
    }
}
export class HostCannotAppendActionToJointPromiseError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super('Cannot append action to joint promise', blockHash, blockHeight);
    }
}
export class HostCannotReturnJointPromiseError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Cannot return joint promise`, blockHash, blockHeight);
    }
}
export class HostInvalidPromiseResultIndexError extends ContractExecutionError {
    constructor(resultIndex, blockHash, blockHeight) {
        super(`Invalid promise result index ${resultIndex}`, blockHash, blockHeight);
        this.resultIndex = resultIndex;
    }
}
export class HostInvalidRegisterIdError extends ContractExecutionError {
    constructor(registerId, blockHash, blockHeight) {
        super(`Invalid register id ${registerId}`, blockHash, blockHeight);
        this.registerId = registerId;
    }
}
export class HostIteratorWasInvalidatedError extends ContractExecutionError {
    constructor(iteratorIndex, blockHash, blockHeight) {
        super(`Iterator ${iteratorIndex} was invalidated`, blockHash, blockHeight);
        this.iteratorIndex = iteratorIndex;
    }
}
export class HostMemoryAccessViolationError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Memory access violation`, blockHash, blockHeight);
    }
}
export class HostInvalidReceiptIndexError extends ContractExecutionError {
    constructor(receiptIndex, blockHash, blockHeight) {
        super(`Invalid receipt index ${receiptIndex}`, blockHash, blockHeight);
        this.receiptIndex = receiptIndex;
    }
}
export class HostInvalidIteratorIndexError extends ContractExecutionError {
    constructor(iteratorIndex, blockHash, blockHeight) {
        super(`Invalid iterator index ${iteratorIndex}`, blockHash, blockHeight);
        this.iteratorIndex = iteratorIndex;
    }
}
export class HostInvalidAccountIdError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Invalid account ID`, blockHash, blockHeight);
    }
}
export class HostInvalidMethodNameError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Invalid method name`, blockHash, blockHeight);
    }
}
export class HostInvalidPublicKeyError extends ContractExecutionError {
    constructor(blockHash, blockHeight) {
        super(`Invalid public key`, blockHash, blockHeight);
    }
}
export class HostProhibitedInViewError extends ContractExecutionError {
    constructor(methodName, blockHash, blockHeight) {
        super(`Method ${methodName} is prohibited in view calls`, blockHash, blockHeight);
        this.methodName = methodName;
    }
}
export class HostNumberOfLogsExceededError extends ContractExecutionError {
    constructor(limit, blockHash, blockHeight) {
        super(`Number of logs exceeded limit ${limit}`, blockHash, blockHeight);
        this.limit = limit;
    }
}
export class HostKeyLengthExceededError extends ContractExecutionError {
    constructor(length, limit, blockHash, blockHeight) {
        super(`Key length ${length} exceeds limit ${limit}`, blockHash, blockHeight);
        this.length = length;
        this.limit = limit;
    }
}
export class HostValueLengthExceededError extends ContractExecutionError {
    constructor(length, limit, blockHash, blockHeight) {
        super(`Value length ${length} exceeds limit ${limit}`, blockHash, blockHeight);
        this.length = length;
        this.limit = limit;
    }
}
export class HostTotalLogLengthExceededError extends ContractExecutionError {
    constructor(length, limit, blockHash, blockHeight) {
        super(`Total log length ${length} exceeds limit ${limit}`, blockHash, blockHeight);
        this.length = length;
        this.limit = limit;
    }
}
export class HostNumberPromisesExceededError extends ContractExecutionError {
    constructor(limit, numberOfPromises, blockHash, blockHeight) {
        super(`Number of promises ${numberOfPromises} exceeds limit ${limit}`, blockHash, blockHeight);
        this.limit = limit;
        this.numberOfPromises = numberOfPromises;
    }
}
export class HostNumberInputDataDependenciesExceededError extends ContractExecutionError {
    constructor(limit, numberOfInputDataDependencies, blockHash, blockHeight) {
        super(`Number of input data dependencies ${numberOfInputDataDependencies} exceeds limit ${limit}`, blockHash, blockHeight);
        this.limit = limit;
        this.numberOfInputDataDependencies = numberOfInputDataDependencies;
    }
}
export class HostReturnedValueLengthExceededError extends ContractExecutionError {
    constructor(length, limit, blockHash, blockHeight) {
        super(`Returned value length ${length} exceeds limit ${limit}`, blockHash, blockHeight);
        this.length = length;
        this.limit = limit;
    }
}
export class HostSizeExceededError extends ContractExecutionError {
    constructor(size, limit, blockHash, blockHeight) {
        super(`Contract size ${size} exceeds limit ${limit}`, blockHash, blockHeight);
        this.size = size;
        this.limit = limit;
    }
}
export class HostDeprecatedError extends ContractExecutionError {
    constructor(methodName, blockHash, blockHeight) {
        super(`Method ${methodName} is deprecated`, blockHash, blockHeight);
        this.methodName = methodName;
    }
}
export class HostECRecoverError extends ContractExecutionError {
    constructor(message, blockHash, blockHeight) {
        super(`ECRecover error: ${message}`, blockHash, blockHeight);
    }
}
export class HostAltBn128InvalidInputError extends ContractExecutionError {
    constructor(message, blockHash, blockHeight) {
        super(`AltBn128 invalid input: ${message}`, blockHash, blockHeight);
    }
}
export class HostEd25519VerifyInvalidInputError extends ContractExecutionError {
    constructor(message, blockHash, blockHeight) {
        super(`Ed25519 verify invalid input: ${message}`, blockHash, blockHeight);
    }
}
