// near-api-js/lib/transactions/actions.js
import { Enum } from '../types/index.js';
export class FunctionCallPermission {
    constructor({ allowance, receiverId, methodNames, }) {
        this.allowance = allowance;
        this.receiverId = receiverId;
        this.methodNames = methodNames;
    }
}
export class FullAccessPermission {
}
export class AccessKeyPermission extends Enum {
    constructor(props) {
        super(props);
        for (const [k, v] of Object.entries(props || {})) {
            this[k] = v;
            this.enum = k;
        }
    }
}
export class AccessKey {
    constructor({ nonce, permission }) {
        this.nonce = nonce;
        this.permission = permission;
    }
}
export class CreateAccount {
}
export class DeployContract {
    constructor({ code }) {
        this.code = code;
    }
}
export class FunctionCall {
    constructor({ methodName, args, gas, deposit, }) {
        this.methodName = methodName;
        this.args = args;
        this.gas = gas;
        this.deposit = deposit;
    }
}
export class GlobalContractDeployMode extends Enum {
    constructor(props) {
        super(props);
        for (const [k, v] of Object.entries(props || {})) {
            this[k] = v;
            this.enum = k;
        }
    }
}
export class GlobalContractIdentifier extends Enum {
    constructor(props) {
        super(props);
        for (const [k, v] of Object.entries(props || {})) {
            this[k] = v;
            this.enum = k;
        }
    }
}
export class DeployGlobalContract {
    constructor({ code, deployMode }) {
        this.code = code;
        this.deployMode = deployMode;
    }
}
export class UseGlobalContract {
    constructor({ contractIdentifier }) {
        this.contractIdentifier = contractIdentifier;
    }
}
export class Transfer {
    constructor({ deposit }) {
        this.deposit = deposit;
    }
}
export class Stake {
    constructor({ stake, publicKey }) {
        this.stake = stake;
        this.publicKey = publicKey;
    }
}
export class AddKey {
    constructor({ publicKey, accessKey }) {
        this.publicKey = publicKey;
        this.accessKey = accessKey;
    }
}
export class DeleteKey {
    constructor({ publicKey }) {
        this.publicKey = publicKey;
    }
}
export class DeleteAccount {
    constructor({ beneficiaryId }) {
        this.beneficiaryId = beneficiaryId;
    }
}
export class SignedDelegate {
    constructor({ delegateAction, signature }) {
        this.delegateAction = delegateAction;
        this.signature = signature;
    }
}
/**
 * Contains a list of the valid transaction Actions available with this API
 * @see {@link https://nomicon.io/RuntimeSpec/Actions.html | Actions Spec}
 */
export class Action extends Enum {
    constructor(props) {
        super(props);
        for (const [k, v] of Object.entries(props || {})) {
            this[k] = v;
            this.enum = k;
        }
    }
}
