// near-api-js/lib/transactions/signature.js
import { KeyType } from '../crypto/index.js';
import { Enum } from '../types/index.js';
class ED25519Signature {
    constructor() {
        this.keyType = KeyType.ED25519;
    }
}
class SECP256K1Signature {
    constructor() {
        this.keyType = KeyType.SECP256K1;
    }
}
function resolveEnumKeyName(keyType) {
    switch (keyType) {
        case KeyType.ED25519: {
            return 'ed25519Signature';
        }
        case KeyType.SECP256K1: {
            return 'secp256k1Signature';
        }
        default: {
            throw Error(`unknown type ${keyType}`);
        }
    }
}
export class Signature extends Enum {
    constructor(signature) {
        const keyName = resolveEnumKeyName(signature.keyType);
        super({ [keyName]: signature });
        this[keyName] = signature;
        this.enum = keyName;
    }
    get signature() {
        return this.ed25519Signature || this.secp256k1Signature;
    }
    get signatureType() {
        if (typeof this.signature === 'undefined')
            throw new Error(`Property 'signature' of Signature is undefined`);
        return this.signature.keyType;
    }
    get data() {
        if (typeof this.signature === 'undefined')
            throw new Error(`Property 'signature' of Signature is undefined`);
        return this.signature.data;
    }
}
