// near-api-js/lib/providers/fetch_json.js
import { TypedError } from '../types/index.js';
const BACKOFF_MULTIPLIER = 1.5;
const RETRY_NUMBER = 10;
const RETRY_DELAY = 0;
/**
 * Simple exponential backoff implementation.
 * Retries a function with exponentially increasing delays.
 */
async function backOff(fn, options) {
    const { numOfAttempts, timeMultiple, startingDelay, retry } = options;
    let delay = startingDelay;
    for (let attempt = 1; attempt <= numOfAttempts; attempt++) {
        try {
            return await fn();
        }
        catch (error) {
            if (attempt === numOfAttempts || !retry(error)) {
                throw error;
            }
            if (delay > 0) {
                await new Promise((resolve) => setTimeout(resolve, delay));
            }
            delay = delay === 0 ? 1 : delay * timeMultiple;
        }
    }
    return undefined;
}
export function retryConfig(numOfAttempts = RETRY_NUMBER, timeMultiple = BACKOFF_MULTIPLIER, startingDelay = RETRY_DELAY) {
    return {
        numOfAttempts: numOfAttempts,
        timeMultiple: timeMultiple,
        startingDelay: startingDelay,
        retry: (e) => {
            if ([503, 500, 408].includes(e.cause)) {
                return true;
            }
            if (e.toString().includes('FetchError') || e.toString().includes('Failed to fetch')) {
                return true;
            }
            return false;
        },
    };
}
export class ProviderError extends Error {
    constructor(message, options) {
        super(message, options);
        this.cause = -1;
        if (options.cause) {
            this.cause = options.cause;
        }
    }
}
/**
 * Performs an HTTP request to an RPC endpoint
 * @param url URL for the HTTP request
 * @param json Request body
 * @param headers HTTP headers to include with the request
 * @returns Promise<any> }arsed JSON response from the HTTP request.
 */
export async function fetchJsonRpc(url, json, headers, retryConfig) {
    const response = await backOff(async () => {
        const res = await fetch(url, {
            method: 'POST',
            body: JSON.stringify(json),
            headers: { ...headers, 'Content-Type': 'application/json' },
        });
        const { status } = res;
        if (status === 500) {
            throw new ProviderError('Internal server error', { cause: status });
        }
        else if (status === 408) {
            throw new ProviderError('Timeout error', { cause: status });
        }
        else if (status === 429) {
            throw new ProviderError(await res.text(), { cause: status });
        }
        else if (status === 502) {
            throw new ProviderError(`${url} Bad Gateway`, { cause: status });
        }
        else if (status === 503) {
            throw new ProviderError(`${url} unavailable`, { cause: status });
        }
        return res;
    }, retryConfig);
    if (!response) {
        throw new TypedError(`Exceeded ${RETRY_NUMBER} attempts for ${url}.`, 'RetriesExceeded');
    }
    return await response.json();
}
